File: api/src/routes/pages.ts
import { FastifyInstance } from 'fastify';
import pool from '../db/pool.js';
import minio from '../minio.js';

export function registerPageRoutes(app: FastifyInstance): void {
  app.get('/api/pages/:id/screenshots', async (request, reply) => {
    const { id } = request.params as any;
    const result = await pool.query(
      'SELECT id, ts, trigger, format, width, height, file_size, object_key FROM screenshots WHERE page_id = $1 ORDER BY ts',
      [id],
    );
    return { pageId: id, screenshots: result.rows };
  });

  app.get('/api/pages/:id/snapshot', async (request, reply) => {
    const { id } = request.params as any;

    const page = await pool.query('SELECT session_id FROM pages WHERE id = $1', [id]);
    if (page.rows.length === 0) return reply.status(404).send({ error: 'Page not found' });

    try {
      const stream = await minio.getObject('obs-raw', `sessions/${page.rows[0].session_id}/rendered.rrweb.zst`);
      reply.header('Content-Type', 'application/json');
      return stream;
    } catch {
      return reply.status(404).send({ error: 'Snapshot not found' });
    }
  });

  app.get('/api/pages/:id/screenshot/:screenshotId', async (request, reply) => {
    const { id: pageId, screenshotId } = request.params as any;
    const result = await pool.query('SELECT object_key, format FROM screenshots WHERE id = $1 AND page_id = $2', [screenshotId, pageId]);
    if (result.rows.length === 0) return reply.status(404).send({ error: 'Screenshot not found' });

    try {
      const stream = await minio.getObject('obs-raw', result.rows[0].object_key);
      reply.header('Content-Type', `image/${result.rows[0].format}`);
      return stream;
    } catch {
      return reply.status(404).send({ error: 'Screenshot file not found in storage' });
    }
  });
}

File: api-rust/src/routes/pages.rs
use axum::extract::{Path, State};
use axum::Json;
use serde::Serialize;
use sqlx::PgPool;
use std::sync::Arc;
use uuid::Uuid;

use crate::db::schema::Screenshot;
use crate::AppState;

#[derive(Debug, Serialize)]
pub struct ScreenshotListResponse {
    pub page_id: String,
    pub screenshots: Vec<Screenshot>,
}

pub async fn list_screenshots(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Json<ScreenshotListResponse> {
    let pool = &state.db_pool;
    let screenshots: Vec<Screenshot> = sqlx::query_as(
        "SELECT * FROM screenshots WHERE page_id = $1 ORDER BY ts DESC"
    )
    .bind(id)
    .fetch_all(pool)
    .await
    .unwrap_or_default();

    Json(ScreenshotListResponse {
        page_id: id.to_string(),
        screenshots,
    })
}

pub async fn get_snapshot(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<(axum::http::StatusCode, [(axum::http::HeaderName, &'static str); 1], axum::body::Body), axum::http::StatusCode> {
    let key = format!("sessions/{}/rendered.rrweb.zst", id);
    match state.minio.get_object("obs-raw", &key).await {
        Ok(data) => Ok((
            axum::http::StatusCode::OK,
            [(axum::http::header::CONTENT_TYPE, "application/json")],
            axum::body::Body::from(data),
        )),
        Err(_) => Err(axum::http::StatusCode::NOT_FOUND),
    }
}

pub async fn get_screenshot(
    State(state): State<Arc<AppState>>,
    Path((page_id, screenshot_id)): Path<(Uuid, Uuid)>,
) -> Result<(axum::http::StatusCode, [(&'static str, &'static str); 0], axum::body::Body), axum::http::StatusCode> {
    let pool = &state.db_pool;
    let screenshot: Option<Screenshot> = sqlx::query_as(
        "SELECT * FROM screenshots WHERE id = $1 AND page_id = $2"
    )
    .bind(screenshot_id)
    .bind(page_id)
    .fetch_optional(pool)
    .await
    .unwrap_or(None);

    match screenshot {
        Some(s) => {
            match state.minio.get_object("obs-raw", &s.object_key).await {
                Ok(data) => {
                    Ok((axum::http::StatusCode::OK, [], axum::body::Body::from(data)))
                }
                Err(_) => Err(axum::http::StatusCode::NOT_FOUND),
            }
        }
        None => Err(axum::http::StatusCode::NOT_FOUND),
    }
}
